### ISSUE TITLE
Some benchmarks are very noisy

### ISSUE THREAD

**Issue by sjakobi** [haskell-unordered-containers/unordered-containers#332]
These are comparisons between runs on the same code:

![image](https://user-images.githubusercontent.com/3664523/143896800-3c13232a-4349-49e4-8477-7db3792213ee.png)

![image](https://user-images.githubusercontent.com/3664523/143897130-63856569-1f7d-4363-a6c1-09f9201d28dc.png)

(Let me know if you need a text version instead of these screenshots.)

I did try to make my machine pretty quiet and also used `"-fproc-alignment=64" "-with-rtsopts=-A32m"` to improve reliability. I don't know why these benchmarks are still so very noisy, but I note that most of these are on the slower end of our benchmark suite.

It also seems noteworthy that hardly any of the `containers` and `hashmap` benchmarks are included, apparently more than would be explained by their smaller share of the suite.

Maybe implementing https://github.com/haskell-unordered-containers/unordered-containers/issues/293 would help?!

**Comment by doyougnu** [comment]
I started having a go at #17 and got distracted by this. I don't have conclusive evidence but I suspect that a lot of this noise is due to stalled backend cycles in the CPU. This is observable with `perf`:

Here I run the benchmarks just for lookups
```
[nix-shell:~/programming/haskell/unordered-containers]$ perf stat -r 5 ./dist-newstyle/build/x86_64-linux/ghc-8.10.7/unordered-containers-0.2.16.0/b/benchmarks/build/benchmarks/benchmarks -m prefix "HashMap/lookup"
benchmarked HashMap/lookup/String
time                 1.585 ms   (1.579 ms .. 1.596 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.608 ms   (1.604 ms .. 1.612 ms)
std dev              10.69 μs   (8.154 μs .. 15.22 μs)

...

benchmarked HashMap/lookup-miss/Int
time                 268.4 μs   (262.2 μs .. 275.5 μs)
                     0.998 R²   (0.994 R² .. 1.000 R²)
mean                 261.8 μs   (260.6 μs .. 262.9 μs)
std dev              1.956 μs   (1.459 μs .. 2.744 μs)


 Performance counter stats for './dist-newstyle/build/x86_64-linux/ghc-8.10.7/unordered-containers-0.2.16.0/b/benchmarks/build/benchmarks/benchmarks -m prefix HashMap/lookup' (5 runs):

         32,708.84 msec task-clock:u              #    1.000 CPUs utilized            ( +-  0.27% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
             8,390      page-faults:u             #    0.257 K/sec                    ( +-  0.01% )
   129,934,644,209      cycles:u                  #    3.972 GHz                      ( +-  0.26% )
     3,223,693,319      stalled-cycles-frontend:u #    2.48% frontend cycles idle     ( +-  4.37% )
    42,343,991,912      stalled-cycles-backend:u  #   32.59% backend cycles idle      ( +-  0.35% )
   207,466,788,278      instructions:u            #    1.60  insn per cycle
                                                  #    0.20  stalled cycles per insn  ( +-  0.58% )
    36,443,482,339      branches:u                # 1114.178 M/sec                    ( +-  0.57% )
       347,372,412      branch-misses:u           #    0.95% of all branches          ( +-  0.44% )

           32.7236 +- 0.0878 seconds time elapsed  ( +-  0.27% )
```

`32%` is quite a bit of stalled cycles. Typically this means that on my CPU (AMD Ryzen 2700X) the cpu spends a lot of time waiting for memory to load into the CPU caches.

So I wonder if some `prefetchValue3#`s  from `GHC.Exts` would reduce noise _and_ improve performance. For example, I think in `lookupInArrayCont` we could try to prefetch several array indices at a time. Deciding on the right number will require some investigation but if successful we'll be saving millions stalled CPU Cycles.

**Comment by treeowl** [comment]
@sjakobi yes, for accessibility purposes, please use text rather than images of text. Images should only be used if they add substantial value (e.g., diagrams or graphs).

**Comment by sjakobi** [comment]
@treeowl Alright. I've updated the issue description. Luckily, `tasty-bench` makes it pretty easy to get text-based output regarding differences between benchmark runs.

**Comment by treeowl** [comment]
Thanks!

**Comment by sjakobi** [comment]
I have updated the issue description again to include benchmark results from `containers` and `hashmap`.

**Comment by sjakobi** [comment]
@doyougnu That's a very interesting lead! An array-related underlying problem would explain why the benchmarks for `containers` and `hashmap` (which is based on `IntMap`) are less noisy than those for `u-c`.

It would be great if you could take a stab at fixing this! I've got to admit that this is the first time I hear of `prefetchValue3#`, so this is quite a bit out of my usual comfort zone.

**Comment by sjakobi** [comment]
For reference:

* I've found https://gitlab.haskell.org/ghc/ghc/-/issues/10809 which requests dedicated prefetch primops for `Small[Mutable]Array`.

* Some documentation for the existing prefetch primops can be found in [`primops.txt.pp`](https://gitlab.haskell.org/ghc/ghc/-/blob/9552781abf3d199d01b01ed4b7d04abdc54b6b26/compiler/GHC/Builtin/primops.txt.pp#L3838-3886).

**Comment by doyougnu** [comment]
@sjakobi I've got an MR up for `10809` now: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7245
